torch/autograd/functional.py

```diff
@@ -0,0 +1,717 @@
+    # strict and create graph allow us to detect when it is appropriate to raise an error
+    # stage gives us information of which backward call we consider to give good error message
+    if stage not in ["back", "back_trick", "double_back", "double_back_trick"]:
+        raise RuntimeError("Invalid stage argument '{}' to _fill_in_zeros".format(stage))
+
+    res = tuple()
+    for i, grads_i in enumerate(grads):
+        if grads_i is None:
+            if strict:
+                if stage == "back":
+                    raise RuntimeError("The output of the user-provided function is independent of "
+                                       "input {}. This is not allowed in strict mode.".format(i))
+                elif stage == "back_trick":
+                    raise RuntimeError("The gradient with respect to the input is independent of entry {}"
+                                       " in the grad_outputs when using the double backward trick to compute"
+                                       " forward mode gradients. This is not allowed in strict mode.".format(i))
+                elif stage == "double_back":
+                    raise RuntimeError("The jacobian of the user-provided function is independent of "
+                                       "input {}. This is not allowed in strict mode.".format(i))
+                else:
+                    raise RuntimeError("The hessian of the user-provided function is independent of "
+                                       "entry {} in the grad_jacobian. This is not allowed in strict "
+                                       "mode as it prevents from using the double backward trick to "
+                                       "replace forward mode AD.".format(i))
+
+            grads_i = torch.zeros_like(refs[i])
+        else:
+            if strict and create_graph and not grads_i.requires_grad:
+                if "double" not in stage:
+                    raise RuntimeError("The jacobian of the user-provided function is independent of "
+                                       "input {}. This is not allowed in strict mode when create_graph=True.".format(i))
+                else:
+                    raise RuntimeError("The hessian of the user-provided function is independent of "
+                                       "input {}. This is not allowed in strict mode when create_graph=True.".format(i))
+
+        res += (grads_i,)
+
+    return res
+
+# Public API
+
+def vjp(func, inputs, v=None, create_graph=False, strict=False):
+    r"""Function that computes the dot product between a vector ``v`` and the Jacobian of
+    the given function at the point given by the inputs.
+
+    Args:
+        func (function): a Python function that takes Tensor inputs and returns
+            a tuple of Tensor or a Tensor.
+        inputs (tuple of Tensor or Tensor): inputs to the function ``func``.
+        v (tuple of Tensor or Tensor): The vector that will multiply the Jacobian. Must be the
```

Comment:
nit: this seems to suggest you actually multiply the Jacobian, but that isn't true, right?

Reply:
Multiply in the matrix sense. I could say "The vector used in the product with the Jacobian" but it doesn't feel much better. You think I should change it?

Reply:
I'd put something like "the vector for which the vector jacobian product is computed."  That way you don't say how it is used, just that a vector jacobian product is computed.